resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java

```diff
@@ -889,14 +889,56 @@ protected void restoreMarkers(IResource resource, boolean generateDeltas, IProgr
 			return;
 		}
 		IProject[] projects = ((IWorkspaceRoot) resource).getProjects(IContainer.INCLUDE_HIDDEN);
-		for (IProject project : projects)
-			if (project.isAccessible())
-				markerManager.restore(project, generateDeltas, monitor);
+		runParallel(projects, monitor, project -> {
+			if (project.isAccessible()) {
+				markerManager.restore(project, generateDeltas, null);
+			}
+		});
 		if (Policy.DEBUG_RESTORE_MARKERS) {
 			Policy.debug("Restore Markers for workspace: " + (System.currentTimeMillis() - start) + "ms"); //$NON-NLS-1$ //$NON-NLS-2$
 		}
 	}
 
+	@FunctionalInterface
+	public interface CoreConsumer<T> {
+		/**
+		 * Performs this operation on the given argument.
+		 *
+		 * @param t the input argument
+		 */
+		void accept(T t) throws CoreException;
+	}
+
+	void runParallel(IProject[] projects, IProgressMonitor m, CoreConsumer<IProject> consumer) throws CoreException {
+		// Never use a shared ForkJoinPool.commonPool() as it may be busy with other tasks, which might deadlock.
+		// Also use a custom ForkJoinWorkerThreadFactory, to prevent issues with a
+		// potential SecurityManager, since the threads created by it get no permissions.
+		// See https://github.com/eclipse-platform/eclipse.platform/issues/294
+		ForkJoinPool forkJoinPool = new ForkJoinPool(ForkJoinPool.getCommonPoolParallelism(),
```

Comment:
```suggestion
		ForkJoinPool forkJoinPool = Executors.newWorkStealingPool();
```

Reply:
why? i don't know the difference but the comment above explicitly states a custom ForkJoinWorkerThreadFactory had to be used.

Reply:
I can only see this bug reported against java 8 but we are using java 17 now, if it is still be present, it might be better to have a common method e.g. in `Platform` instead of implementing this pattern over and over again.

Reply:
Changing that is out of scope of this PR.

Reply:
> Changing that is out of scope of this PR.

You are changing it with this PR so please address it here as it is clearly in scope...

Reply:
No way. The custom ForkJoinWorkerThreadFactory was added because someone else had a serious problem not long ago and I am not about to change it on the fly.

Reply:
You don't *change* anything you _add new_ code here. And copy the same problematic code over and over again does not help in fixing the issue, especially if its only because "someone has had a problem but we don't know if its still a problem".

Reply:
> This PR does not add that code here or copies it around. It only moves the existing code into a method.

Okay, then still it looks strange why this is not a general purpose method somewhere in platform ...